README.md

```diff
@@ -2,54 +2,113 @@
+[Metalsmith deployment](https://docs.openstack.org/metalsmith/latest/), where a
+deployment consumes a node.  This allows a user to manage their hardware outside of
+terraform -- perhaps through Ironic's discovery mechanism.
+
+## Nodes
 
+A node describes a hardware resource. If a node of the same name already exists, or
+`node_uuid` is specified, it's details would be updated. This could be used to provide
+BMC information for a discovered node.
+
+Destruction of a node object removes the node from Ironic.
+
+
+```
 resource "ironic_node_v1" "openshift-master-0" {
   name = "openshift-master-0"
-  target_provision_state = "active"
-  user_data = "${file("master.ign")}"
 
   ports = [
     {
-      "address" = "00:bb:4a:d0:5e:38"
-      "pxe_enabled" = "true"
-    }
+      pxe_enabled = true
+      address     = "00:bb:4a:d0:5e:38"
+    },
   ]
-
-  properties {
-    "local_gb" = "50"
-    "cpu_arch" =  "x86_64"
-  }
-
-  instance_info = {
-    "image_source" = "http://172.22.0.1/images/redhat-coreos-maipo-latest.qcow2"
-    "image_checksum" = "26c53f3beca4e0b02e09d335257826fd"
-    "root_gb" = "25"
-    "root_device" = "/dev/vda"
-  }
-
+  
   driver = "ipmi"
+  
   driver_info {
-			"ipmi_port"=      "6230"
-			"ipmi_username"=  "admin"
-			"ipmi_password"=  "password"
-			"ipmi_address"=   "192.168.111.1"
-			"deploy_kernel"=  "http://172.22.0.1/images/ironic-python-agent.kernel"
-			"deploy_ramdisk"= "http://172.22.0.1/images/ironic-python-agent.initramfs"
+    ipmi_port       =  "6230"
+    ipmi_username   =  "admin"
+    ipmi_password   =  "password"
+    ipmi_address    =  "192.168.111.1"
+    deploy_kernel   =  "http://172.22.0.1/images/ironic-python-agent.kernel"
+    deploy_ramdisk  =  "http://172.22.0.1/images/ironic-python-agent.initramfs"
   }
 }
 ```
 
+## Ports
+
+Ports may be specified as part of the node resource, or as a separate `ironic_port_v1`
+declaration.
+
+
+```
+resource "ironic_port_v1" "openshift-master-0-port-0" {
+  node_uuid   = "${ironic_node_v1.openshift-master-0.id}"
+  pxe_enabled = true
+  address     = "00:bb:4a:d0:5e:38"
+}
+```
+
+## Deployment
+
+A deployment will provision a baremetal node, using the information given in the resource.  The `count` metaparameter
+can be used to deploy multiple nodes.
+
+An available node will be selected based on critera described below, and terraform will drive
+the [Ironic state machine](https://docs.openstack.org/ironic/latest/contributor/states.html) to bring the node to an `active` state.
+
+Destruction of a deployment cleans a node and brings it back to the `available` state.
+
+### Node selection
+
+Users can control node selection with two fields. `candidates` is a list of node UUID's that
+could be used.  If specified, the first available node will be used.  The other mechanism is
+by specifying a list of `capabilities` (TODO: provide example), which will restrict which nodes are considered.
```

Comment:
But we also need traits created in order to use it. I don't see the data structure really disappearing anytime soon. It just can't be expected to be there.

Reply:
We need traits created for sure, but that should not be an excuse to rely on capabilities for anything, except for `boot_mode`/`boot_option` sadness.

Reply:
Do I need to consider capabilities anywhere then for the general use case? For metalkube, is boot_mode / boot_option relevant?

Reply:
I suggest you ignore them for the purpose of scheduling. However, there is still a valid case of passing capabilities (as `map[string]string`) to `instance_info`: it can be used to modify `boot_option` and `boot_mode`.